### PR TITLE
Tensor Metrics 4: Integrate tensor observation into training

### PR DESCRIPTION
**Pull request by philipcmonk**
Integrates configurable tensor metrics into the Megatron training loop, including forward observation, optimizer-backed parameter and wgrad sources, scheduling, timing, and logging.  This does not yet include any backward-pass observations.

This is part 4 of a four-PR stack:

1. #6647: TensorMetric core framework
2. #6648: MCore tensor observation sources
3. #6649: metric definitions
4. **This PR:** training-loop integration

## Configuration

Metrics are configured independently as `NAME:INTERVAL` specifications:

```
--tensor-metrics layer-router-health:50 layer-param-l2:100
```

The observer validates unknown metrics, duplicate names, and invalid intervals during startup.

## Training-loop flow

For an iteration with due forward metrics, the observer installs an MCore tensor-observation scope around forward/backward. Each observed tensor is immediately passed through the metric's local prepare() operation so short-lived raw tensors do not need to survive until the end of the step.

After backward and immediately before optimizer.step(), the observer:

1. obtains finalized parameter and wgrad sources;
2. combines them with prepared forward contributions;
3. runs each due metric through TensorMetricExecutor;
4. writes scalar results to TensorBoard and/or Weights & Biases;
5. falls back to rank-last stdout when no writer is configured.

The integration adds a tensor-metrics timer and NVTX ranges.

## Optimizer sources

The optimizer adapters cover:

- FP32 optimizer storage;
- mixed-precision optimizer storage;
- DistributedOptimizer flat shards;
- layer-wise distributed optimizers;
- chained optimizers;
- ordinary and expert parameters;
- GTP and expert-GTP rematerialized storage;
- source-side removal of GTP padding.

If the Tensor Metrics system moves into core/ at some point, then these should move into an overridable function on the optimizer.

A rank-symmetric optimizer manifest supplies canonical names, logical shapes, dtypes, ownership, and placement metadata. Ranks without a locally owned tensor participate with explicit neutral placeholders, ensuring that every rank issues the same collective sequence.

### CODE REVIEW

**Comment by hexinw-nvidia** [comment]
Hi @philipcmonk, could we fold in a small NVRx fix here, or would you take it as a follow-up?

With tensor_metrics on, the metric iterations are much slower than a normal step. They run inside the region timed by the FT step hang-detection section, so that time trips the step timeout and causes false-positive InJob restarts.

The fix keeps the step tight for normal iterations and times only the tensor-metrics iterations under a separate tensor_metrics section with its own larger budget.  Details: https://github.com/NVIDIA-NeMo/nv-mistralai-megatron/pull/458

**Comment by philipcmonk** [comment]
Yes, that makes sense. Let's fold it into this PR.

________________________________

**Comment by hexinw-nvidia** [comment]
Thanks @philipcmonk! I put the NVRx step-timeout fix up as a small PR against your branch: philipcmonk/Megatron-LM#1 (base `pmonk/tensor-metrics-training-integration`). Merging it there folds it into this PR.

**Comment by philipcmonk** [comment]
Thanks!  That looks good, I just merged it into this.
